src/Nethermind/Nethermind.Synchronization.Test/AllocationStrategies/BySpeedStrategyTests.cs

```diff
@@ -62,7 +62,89 @@ public void TestShouldSelectHighestSpeed(int? currentPeerIdx, decimal minDiffPer
         selectedPeerIdx.Should().Be(expectedSelectedPeerIdx);
     }
 
-    private static PeerInfo CreatePeerInfoWithSpeed(long speed, INodeStatsManager nodeStatsManager)
+    [TestCase(1, 0, 0, false)]
+    [TestCase(0, 1, 0, true)]
+    [TestCase(1, 1, 1, false)]
+    [TestCase(1, 1, 2, true)]
+    [TestCase(10, 10, 2, false)]
+    [TestCase(10, 10, 11, true)]
+    [TestCase(10, 0, 11, false)]
+    public void TestMinimumKnownSpeed(int peerWithKnownSpeed, int peerWithUnknownSpeed, int desiredKnownPeer, bool pickedNewPeer)
+    {
+        long?[] peerSpeeds = Enumerable.Repeat<long?>(100, peerWithKnownSpeed)
+            .Concat(Enumerable.Repeat<long?>(null, peerWithUnknownSpeed))
+            .ToArray();
+
+        INodeStatsManager nodeStatsManager = Substitute.For<INodeStatsManager>();
+
+        List<PeerInfo> peers = new();
+        for (int i = 0; i < peerSpeeds.Length; i++)
+        {
+            PeerInfo pInfo = CreatePeerInfoWithSpeed(peerSpeeds[i], nodeStatsManager);
+            peers.Add(pInfo);
+        }
+
+        BySpeedStrategy strategy = new(TransferSpeedType.Bodies, true, 0, 0, 0, desiredKnownPeer);
+
+        PeerInfo? selectedPeer = strategy.Allocate(null, peers, nodeStatsManager, Build.A.BlockTree().TestObject);
+
+        int selectedPeerIdx = peers.IndexOf(selectedPeer);
+        if (pickedNewPeer)
+        {
+            selectedPeerIdx.Should().Be(peerWithKnownSpeed);
+        }
+        else
+        {
+            selectedPeerIdx.Should().Be(0);
+        }
+    }
+
+    [TestCase(10, 0, 0, 0)]
+    [TestCase(10, 0, 1, 0)]
+    [TestCase(10, 10, 1, 0.5)]
+    [TestCase(10, 10, 0.5, 0.25)]
+    [Retry(3)]
```

Comment:
why do we need retry?

Reply:
The logic is random, already have margin of error, but sometimes it still fails.